sdk/core/src/http_client.rs

```diff
@@ -62,69 +74,126 @@ pub trait HttpClient: Send + Sync + std::fmt::Debug {
+//
+//        Ok(response)
+//    }
+//}
+
+#[cfg(feature = "enable_reqwest")]
 #[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
 #[cfg_attr(not(target_arch = "wasm32"), async_trait)]
-impl HttpClient for hyper::Client<HttpsConnector<hyper::client::HttpConnector>> {
+impl HttpClient for reqwest::Client {
     async fn execute_request(
         &self,
         request: Request<Bytes>,
     ) -> Result<Response<Bytes>, Box<dyn std::error::Error + Sync + Send>> {
-        let mut hyper_request = hyper::Request::builder()
-            .uri(request.uri())
-            .method(request.method());
-
+        let mut reqwest_request = self.request(
+            request.method().clone(),
+            url::Url::parse(&request.uri().to_string()).unwrap(),
+        );
         for header in request.headers() {
-            hyper_request = hyper_request.header(header.0, header.1);
+            reqwest_request = reqwest_request.header(header.0, header.1);
         }
 
-        let hyper_request = hyper_request.body(hyper::Body::from(request.into_body()))?;
+        let reqwest_request = reqwest_request.body(request.into_body()).build()?;
 
-        let hyper_response = self.request(hyper_request).await?;
+        let reqwest_response = self.execute(reqwest_request).await?;
 
         let mut response = Response::builder()
-            .status(hyper_response.status())
-            .version(hyper_response.version());
+            .status(reqwest_response.status())
+            .version(reqwest_response.version());
 
-        for (key, value) in hyper_response.headers() {
+        for (key, value) in reqwest_response.headers() {
             response = response.header(key, value);
         }
 
-        let response = response.body(hyper::body::to_bytes(hyper_response.into_body()).await?)?;
+        let response = response.body(reqwest_response.bytes().await?)?;
 
         Ok(response)
     }
-}
 
-#[cfg(feature = "enable_reqwest")]
-#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
-#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
-impl HttpClient for reqwest::Client {
-    async fn execute_request(
+    async fn execute_request2(
         &self,
-        request: Request<Bytes>,
-    ) -> Result<Response<Bytes>, Box<dyn std::error::Error + Sync + Send>> {
-        let mut reqwest_request =
-            self.request(request.method().clone(), &request.uri().to_string());
+        request: &mut crate::Request,
+    ) -> Result<crate::Response, Box<dyn std::error::Error + Sync + Send>> {
+        let mut reqwest_request = self.request(
+            request.method().clone(),
+            url::Url::parse(&request.uri().to_string()).unwrap(),
+        );
         for header in request.headers() {
             reqwest_request = reqwest_request.header(header.0, header.1);
         }
 
-        let reqwest_request = reqwest_request.body(request.into_body()).build()?;
+        let body = request.take_body();
```

Comment:
If we need to replace the body for all subsequent retries, shouldn't we create a method like `body_cloned` or something instead of `take_body`?

Reply:
Yes definitely. Good point! 👍